File: Hard/0010 - regularexpressionmatching/regularexpressionmatching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        x = len(s)
        y = len(p)

        arr = [None] * (x + 1)
        for i in range(x + 1):
            arr[i] = [False] * (y + 1)

        arr[0][0] = True;                    #  typical empty string

        for i in range(2, y + 1):            #  empty string with a sequence of *chars
            if (p[i - 1] == '*'):
                arr[0][i] = arr[0][i - 2]

        for i in range(1, x + 1):
            for j in range(1, y + 1):
                if (p[j - 1] == '*'):        #  zero or more matches
                    arr[i][j] = arr[i][j - 2] or (arr[i - 1][j] and (s[i - 1] == p[j - 2] or p[j - 2] == '.'))
                else:                        #  simple since there is no *
                    arr[i][j] = arr[i - 1][j - 1] and (s[i - 1] == p[j - 1] or p[j - 1] == '.')

        return arr[x][y]

File: Hard/0010 - regularexpressionmatching/regularexpressionmatching.cpp
#include <string>
#include <vector>

class Solution {
public:
    bool isMatch(std::string s, std::string p) {
        int x = s.length(), y = p.length();
        std::vector<std::vector<bool>> arr(x + 1, std::vector<bool>(y + 1, false));
        
        arr[0][0] = true;                   //  typical empty string

        for (int i = 2; i <= y; i++)        //  empty string with a sequence of *chars
            if (p[i - 1] == '*')
                arr[0][i] = arr[0][i - 2];

        for (int i = 1; i <= x; i++) {
            for (int j = 1; j <= y; j++) {
                if (p[j - 1] == '*')        //  zero or more matches
                    arr[i][j] = arr[i][j - 2] || (arr[i - 1][j] && (s[i - 1] == p[j - 2] || p[j - 2] == '.'));
                else                        //  simple since there is no *
                    arr[i][j] = arr[i - 1][j - 1] && (s[i - 1] == p[j - 1] || p[j - 1] == '.');
            }
        }

        return arr[x][y];
    }
};